### ISSUE TITLE
Access key per device or one session per access key

### ISSUE THREAD

**Issue by baseloo** [Jigsaw-Code/outline-server#546]
Hello All,

is there anyway even third party solution to restrict  access key per device or to have one session per access key .

**Comment by swimmer2910** [comment]
This is the most common feature. I'm waiting too.

**Comment by moeinrahimi** [comment]
this isn't implemented already?! it's a must have

**Comment by fortuna** [comment]
We are working on it. I'm hoping to release a version of this in January.

**Comment by parsalotfy** [comment]
Can't wait for it :heart_eyes:

**Comment by ercxar** [comment]
I know this feature request is complex, but I have a suggestion. @fortuna 
Is it possible to instead add a feature to when a new access key is created we could give it a limited availability, like only valid for 30 mins. After that the device that connected to the server during that 30 minutes window will be authorized to use that access key but any new devices during or after that time frame get automatically rejected.

**Comment by rsun-jhg** [comment]
2 years passed, any progress on it ?

**Comment by Serqati** [comment]
Is it completed?

**Comment by hosseinnoob** [comment]
Whats Going On ?
Im Waiting 4 This Feture Too But Nothing Happen And I Know Some One Make Limit On ip Connection But I Dont Khow How

**Comment by parsalotfy** [comment]
we need this feature sooner than government find a way to block it :|

**Comment by kakajan373** [comment]
Hello, I would like to know when you will add this feature, we have been waiting for a year. Thank you.

**Comment by Serqati** [comment]
@hosseinnoob I know how to limit 1 IP per access key but I want limit 1 connection per access key

**Comment by Serqati** [comment]
@fortuna  You are a big lier ((((:

**Comment by Serqati** [comment]
@fortuna wtf

**Comment by mcharyyev** [comment]
please add this feature

**Comment by krakazyabra** [comment]
The easiest way to restrict multiple using - collect mac of device and put it into key model. client can send this info on first connection.

**Comment by alexlii1971** [comment]
Anyway to limit Bandwidth per key?

**Comment by bemasc** [comment]
@alexlii1971 Yes (see [here](https://www.reddit.com/r/outlinevpn/comments/mbmrvp/perkey_data_limits_now_in_outline_manager/)).  This feature request is instead about restricting each key to a single device, not necessarily limiting the amount of data used by each key.

**Comment by torabkheslat** [comment]
any news guys?

**Comment by fortuna** [comment]
The main development here is the introduction of Dynamic Keys/Online Config: https://www.reddit.com/r/outlinevpn/wiki/index/dynamic_access_keys/

That can enable providers provide a different key in every session. This way sharing a key will effectively break it.

**Comment by whoami-idk** [comment]
is there any release date for this feature ?

**Comment by hamidmayeli** [comment]
> It's out!

Hi @daniellacosse, Where? which version number? I cannot see it!

**Comment by whoami-idk** [comment]
@daniellacosse 
@fortuna 
Using the dynamic access key isn't a good idea to prevent sharing access keys or restrict keys per devices,since user can open dynamic urls as http(s) url as generate static access key based on json output.

**Comment by whoami-idk** [comment]
any news ?

**Comment by Beniamiiin** [comment]
Hey, guys. Could someone tell me how can restrict access key per device via dynamic keys? I've implemented dynamic keys feature for my clients but I don't understand how to restrict access key per one connection?

**Comment by Beniamiiin** [comment]
@daniellacosse thanks for the response. But how does this help me solve the issue of people buying one key and then using it on multiple devices? I thought dynamic keys can help with it.

**Comment by Beniamiiin** [comment]
@daniellacosse I made a mistake when I told you that I got the same information from the both connections. I tried one more time and got the expected result. But this information unfortunately don't help me to restrict of sharing one key or I don't understand something.

**Comment by fortuna** [comment]
@sobhan-m94 , if you implement a web server that creates a new static access key and revokes the previous one every time the dynamic key is requested, then you end up with a session key. Starting a new session cancels the previous one, so sharing would make the key mostly unusable, since you won't be able to have two sessions simultaneously.

**Comment by Beniamiiin** [comment]
Guys thanks for your response. 

@daniellacosse yeah, you are right that the user can not share the static key but can share the dynamic key. Everybody who can have a link to a dynamic key can get a static key. In my case request for a dynamic key makes with userId.

The @fortuna idea's seems good, but I don't understand one thing.

Let's agree that my service support only 2 connections per key.

Making vpn connection from Phone 1.
Check the user's list of created static keys. Create a key and put it on the list. The list count is 1. Return the key to the user.
Connection successfully.

Making VPN connection from Phone 2.
Check the user's list of created static keys. Create a key and put it on the list. The list count is 2. Return the key to the user.
Connection successfully.

Disconnect on Phone 2.
Making a new VPN connection from Phone 2.
Check the user's list of created static keys. Create a key and put it on the list. The list count is 3. Remove the first key from the list and remove it(deactivate) from the outline manager.  Return the key to the user.
Connection successfully.

In this case, connection on Phone 1 will be broken.

Do you have any idea how can I figure it out?

**Comment by cynepmann** [comment]
> @sobhan-m94 , if you implement a web server that creates a new static access key and revokes the previous one every time the dynamic key is requested, then you end up with a session key. Starting a new session cancels the previous one, so sharing would make the key mostly unusable, since you won't be able to have two sessions simultaneously.

Hi, could you, please, explain more in detail how this option can be implemented, especially how to track that the dynamic key was requested?

Thanks a lot in advance!

**Comment by Panuchi** [comment]
any news?

**Comment by khamsolt** [comment]
Any news?

**Comment by SonicNorg** [comment]
> Hey @Beniamiiin - you'd effectively have to write a service that calls our Management API to generate a static key and returns it for each user that connects to your dynamic key service. In that service you'd need to store a mapping of IPs/User Agents to keys generated somewhere to avoid regenerating keys for the same client.
> 
> I know that's a lot of work, but the end result is you now have one key that you can give everyone.

Could you please point me to the documentation or manual which helps to do this? Thank you very much.

**Comment by khamsolt** [comment]
> > Hey @Beniamiiin - you'd effectively have to write a service that calls our Management API to generate a static key and returns it for each user that connects to your dynamic key service. In that service you'd need to store a mapping of IPs/User Agents to keys generated somewhere to avoid regenerating keys for the same client.
> > I know that's a lot of work, but the end result is you now have one key that you can give everyone.
> 
> Could you please point me to the documentation or manual which helps to do this? Thank you very much.

I think that's the [link](https://www.reddit.com/r/outlinevpn/wiki/index/dynamic_access_keys/) you need

**Comment by khamsolt** [comment]
> Dynamic key is accessed for the first time from client #1. The client fingerprint is added to the endpoint and the "lockdown timer" is started.

@daniellacosse 

Hi!

Please tell me what to use as a user or device fingerprint. I just still haven't looked at what parameters different devices (ios/android/desktop) send when getting a key via dynamic link. Thanks!

**Comment by topisun** [comment]
So, till now, no option to set 1 key for 1 client only?

**Comment by Kostya415** [comment]
> Hey, guys. Could someone tell me how can restrict access key per device via dynamic keys? I've implemented dynamic keys feature for my clients but I don't understand how to restrict access key per one connection.

Hi! Did you solve the problem?

**Comment by Kostya415** [comment]
> @sobhan-m94 , if you implement a web server that creates a new static access key and revokes the previous one every time the dynamic key is requested, then you end up with a session key. Starting a new session cancels the previous one, so sharing would make the key mostly unusable, since you won't be able to have two sessions simultaneously.

@fortuna do you have some news about this?

**Comment by Artem-Mit** [comment]
Hey, guys. Any news about this?

**Comment by fortuna** [comment]
Currently, our focus is on adding WebSocket support, so we are unable to prioritize this request at the moment. Additionally, this functionality can be built on top of Outline without requiring any changes to the core product. It’s something a third party could develop and provide to the community.

Our team is concentrating on features that require direct implementation, leaving room for extensions like this to be developed externally.

**Comment by coldenflo** [comment]
hey, any news?

**Comment by Qvvan** [comment]
> Привет, есть новости?

Hello everyone, you don't need to generate keys using Outline Manager. Use 3x-ui instead; it has a configuration option that allows you to set IP restrictions per device. It works with all protocols, including VLESS, Shadowsocks, and others.

**Comment by megam19** [comment]
This is not the best method for 3x-ui. If I have a single router with the provider's IP, 10 users (3x-ui), and a limit of 1 IP, then when 5 hosts connect simultaneously from the local network, 3x-ui (fail2ban) will block my router's IP, and all 10 users (3x-ui) will have their VPN connections stopped.
